### ISSUE TITLE
Support createdump functionality in the app sandbox

### ISSUE THREAD

**Issue by rolfbjarne** [dotnet/runtime#119945]
The 'createdump' tool doesn't currently work in the app sandbox (i.e. when publishing to the App Mac Store).

Ref: https://github.com/xamarin/xamarin-macios/issues/16189

Making it work as a standalone executable has numerous challenges:

* `createdump` has to be signed with a bundle identifier. It doesn't seem like the same bundle identifier as the main app can be used, although I wasn't able to confirm this (in any case it sounds like a bad idea to re-use the bundle identifier).
* `createdump` requires a few custom entitlements: https://github.com/dotnet/runtime/blob/main/eng/pipelines/common/createdump-entitlements.plist
* `createdump` would also need the sandbox entitlement to publish to the App Store (`com.apple.security.app-sandbox`).
* I'm not entirely sure whether a custom provisioning profile is required. If so, `createdump` would have to be put inside an app-like directory structure: https://developer.apple.com/documentation/Xcode/signing-a-daemon-with-a-restricted-entitlement. This is inconvenient, because CoreCLR expects `createdump` next to one of its dylibs. Maybe this can be worked around with a symlink?
* It seems that since a bundle identifier is required, an Info.plist must be embedded inside the executable, which is done when the executable is linked (by passing `-sectcreate __TEXT __info_plist path/to/Info.plist` to ld)... but we get the executable already linked from the runtime pack.

Also interesting: https://developer.apple.com/documentation/xcode/embedding-a-helper-tool-in-a-sandboxed-app

An alternative approach seems to be to embed the `createdump` functionality as a static library, and then potentially fork before generating the dump. This is apparently what .NET does for single-file executables.

Ref: https://github.com/dotnet/runtime/pull/84864
Ref: https://github.com/dotnet/runtime/pull/89203
Ref: https://discord.com/channels/732297728826277939/732297808148824115/1149344656480550912

In any case, we don't even know if there's community interest in supporting createdump functionality in the app sandbox, so if you're interested, please upvote this issue. If we get enough interest, we'll look into it, otherwise we'll eventually close this issue.

**Comment by rolfbjarne** [comment]
Fixing dotnet/macios#23848 would probably make this trivial to fix.

**Comment by steveisok** [comment]
Can you transfer this over to runtime? We're going to look at this generally as part of consolidation.

**Comment by rolfbjarne** [comment]
@steveisok note that this is only applicable to macOS and Mac Catalyst apps, iOS and tvOS apps aren't allowed to execute other executables.

**Comment by steveisok** [comment]
@dotnet/dotnet-diag as I mentioned earlier, we should factor this in when planning diagnostic work for runtime consolidation in 11. It's very likely once we decide on the options, we come back to this issue and fold it in.

**Comment by Youssef1313** [comment]
https://github.com/Youssef1313/DumpTestMacOS/ This repo shows also some issues with creating dumps on macOS. For now only workaround we have in the testing space is disabling apphost, and running directly under dotnet instead. The GitHub Actions in that repo sets up different combinations.

I hope a fix for this issue would also fix the issues we see in the testing space.

Note that xunit.v3 doesn't set allow disabling apphosts as of today. So the workaround isn't available for all xunit.v3 users. https://github.com/xunit/xunit/issues/3432

**Comment by rolfbjarne** [comment]
While not a direct fix for this issue, #131410 would serve the same purpose (allowing crash reports for sandboxed apps), but using the inproc crash reporter.

**Comment by lateralusX** [comment]
@rolfbjarne, given that we land https://github.com/dotnet/runtime/pull/131410, do you believe it's good enough to close out this issue?

**Comment by rolfbjarne** [comment]
> [@rolfbjarne](https://github.com/rolfbjarne), given that we land [#131410](https://github.com/dotnet/runtime/pull/131410), do you believe it's good enough to close out this issue?

Yes, I think we can close this issue then. I can always reopen if I change my mind.
